src/stages/stage5/timeline.c

```diff
@@ -37,537 +27,854 @@ static void stage5_dialog_post_midboss(void) {
+	int difficulty = difficulty_value(8, 6, 4, 3);
 
+	for (int x = 0; x < 600; x++) {
+		PROJECTILE(
+			.proto = pp_rice,
+			.pos = e->pos + 20 * cdir(M_TAU * rng_real()),
+			.color = RGB(0, 0, 255),
+			.move = move_linear(cdir(M_TAU  * rng_real())),
+		);
+		play_sfx_ex("shot3", 0, false);
+		WAIT(difficulty);
 	}
+}
 
-	AT(5390) {
-		create_enemy1c(30.0*I, 1500, Fairy, stage5_superbullet, 2 + I);
-	}
+TASK(miner_swirls_3, {
+	int num;
+	cmplx start;
+	cmplx velocity;
+}) {
+	for (int x = 0; x < ARGS.num; x++) {
+		INVOKE_TASK(miner_swirl, {
+			.start = 32 + (VIEWPORT_W - 2 * 32) * (x % 2) + VIEWPORT_H * I,
+			.velocity = ARGS.velocity,
+		});
 
-	AT(5500) {
-		create_enemy1c(VIEWPORT_W+20 + VIEWPORT_H*0.6*I, 2000, BigFairy, stage5_lightburst, -2*I - 2);
-		create_enemy1c(			 -20 + VIEWPORT_H*0.6*I, 2000, BigFairy, stage5_lightburst, -2*I + 2);
+		WAIT(60);
 	}
+}
 
-	AT(5600) {
-		enemy_kill_all(&global.enemies);
+TASK(miner_swirls_2, {
+	int num;
+	int time;
+	cmplx start;
+	cmplx velocity;
+}) {
+	int delay = ARGS.time / ARGS.num;
+	for (int x = 0; x < ARGS.num; x++) {
+		INVOKE_TASK(miner_swirl, {
+			.start = ARGS.start * x,
+			.velocity = ARGS.velocity,
+		});
+
+		WAIT(delay);
 	}
+}
 
-	{
-		int cnt = 5;
-		int step = 10;
-		double ofs = 42*2;
-
-		FROM_TO(5620, 5620 + step*cnt-1, step) {
-			cmplx src1 = -ofs/4               + (-ofs/4 +      _i    * (VIEWPORT_H-2*ofs)/(cnt-1))*I;
-			cmplx src2 = (VIEWPORT_W + ofs/4) + (-ofs/4 + (cnt-_i-1) * (VIEWPORT_H-2*ofs)/(cnt-1))*I;
-			cmplx dst1 = ofs                  + ( ofs   +      _i    * (VIEWPORT_H-2*ofs)/(cnt-1))*I;
-			cmplx dst2 = (VIEWPORT_W - ofs)   + ( ofs   + (cnt-_i-1) * (VIEWPORT_H-2*ofs)/(cnt-1))*I;
-
-			create_enemy2c(src1, 2000, Swirl, stage5_magnetto, dst1, dst2);
-			create_enemy2c(src2, 2000, Swirl, stage5_magnetto, dst2, dst1);
-		}
+TASK(miner_swirls_1, {
+	int num;
+	int time;
+	cmplx start;
+	cmplx velocity;
+}) {
+	int delay = ARGS.time / ARGS.num;
+	for (int x = 0; x < ARGS.num; x++) {
+		INVOKE_TASK(miner_swirl, {
+			.start = VIEWPORT_W + 200.0 * I * rng_real(),
+			.velocity = ARGS.velocity,
+		});
+		WAIT(delay);
 	}
+}
 
-	FROM_TO(6100, 6350, 60-12*global.diff) {
-		tsrand_fill(2);
-		create_enemy1c(VIEWPORT_W+200.0*I*afrand(0), 500, Swirl, stage5_miner, -3+2.0*I*afrand(1));
-	}
+TASK(superbullet_fairy, {
+	cmplx pos;
+	cmplx acceleration;
+	real offset;
+}) {
+	Enemy *e = TASK_BIND(espawn_fairy_blue(ARGS.pos, ITEMS(.points = 4, .power = 2)));
 
-	FROM_TO(6300, 6350, 50) {
-		create_enemy1c(VIEWPORT_W/4+VIEWPORT_W/2*!(_i&1), 2000, BigFairy, stage5_limiter, 2*I);
-	}
+	e->move = move_towards(e->pos + ARGS.acceleration * 70 + ARGS.offset, 0.05);
+	WAIT(60);
 
-	AT(6960) {
-		stage_unlock_bgm("stage5");
-		global.boss = stage5_spawn_boss();
+	int difficulty = difficulty_value(1, 2, 3, 4);
+	for (int x = 0; x < 140; x++) {
+		cmplx n = cdir(M_PI * sin(x / (8.0 + difficulty) + rng_real() * 0.1) * carg(global.plr.pos - e->pos));
```

Comment:
Looks suspiciously wrong. The original line was this, I think?
```c
cmplx n = cexp(I*M_PI*sin(_i/(8.0+global.diff)+frand()*0.1)+I*carg(global.plr.pos-e->pos));
```

In that case it is almost correct, except instead of multiplying by the `carg(...)` you have to add it, like the original does... and then you preferably get rid of the `carg`. I explain that part in detail below.

The expression is confusing so let's break it down.

**Step 0**: de-laochainlanize the code:
```c
cmplx n = cexp(I * M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1) + I * carg(global.plr.pos - e->pos));
```

**Step 1**: replace `cexp(I * x)` with `cdir(x)`. There are 2 `I`s here, so let's factor them out:
```c
cmplx n = cexp(I * (M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1) + carg(global.plr.pos - e->pos)));
```
Then we can substitute `cdir()` easily:
```c
cmplx n = cdir(M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1) + carg(global.plr.pos - e->pos));
```

**Step 2**: get rid of that `carg`:

What we have here is a sum of two angles, `A` and `B`
```c
real A = M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1);
real B = carg(global.plr.pos - e->pos);
cmplx n = cdir(A + B);
```

Since complex multiplication is equivalent to a scale and a rotation, we can rewrite `n` as follows:
```c
cmplx n = cdir(A) * cdir(B);
```

If we represent A and B as direction vectors rather than angles, we get:
```c
cmplx A = cdir(M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1));
cmplx B = cdir(carg(global.plr.pos - e->pos));
cmplx n = A * B;
```

Let's focus on B now:
```c
cdir(carg(global.plr.pos - e->pos))
```
This expression basically says "get the angle (i.e. argument) of complex `global.plr.pos - e->pos`, and make another complex with the same angle but magnitude = 1". That's a very roundabout way of spelling `cnormalize(global.plr.pos - e->pos)`.
It's also mathematically equivalent to `(global.plr.pos - e->pos) / cabs(global.plr.pos - e->pos)` (`cnormalize` does a slightly optimized version of that).

So:
```c
cmplx A = cdir(M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1));
cmplx B = cnormalize(global.plr.pos - e->pos);
cmplx n = A * B;
```

Folding A and B back into n:
```c
cmplx n = cdir(M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1)) * cnormalize(global.plr.pos - e->pos);
```

### TL;DR

To deal with such expressions you just need to know the following identities:

```
cdir(A + B)  ==  cdir(A) * cdir(B)
```
```
cdir(carg(X))  ==  cnormalize(X)
```
And from these follows:
```
cdir(A + carg(B))  ==  cdir(A) * cnormalize(B)
```